compiler/identify_token.cpp

```diff
@@ -27,120 +28,140 @@ void init_isIdChar() {
   isIdChar['_'] = 1;
 }
 
+/*
+    The function intializes the isBlankChar array. If a character c:
+    ---> is newline, space or tab character, then isBlankChar[c]=1
+    ---> Otherwise, isBlankChar[c]=0
+*/
+void init_isBlankChar() {
+  isBlankChar[' '] = 1;
+  isBlankChar['\t'] = 1;
+  isBlankChar['\n'] = 1;
+}
```

Comment:
It's not worth it. :) It's just 3 tests worst case with the function potentially inlined by the compiler with the branch predictor doing magic as you will have consecutive white space.

Reply:
I used this function to initialize the isBlankChar array, in the same way I initialized the isIdChar array. The information whether a character is blank will be later provided by the character map, so I though I use the information like this for now and just change it later. 
Is the solution you are proposing having a function 
```
bool isBlankChar(char c) {return input==' ' || input=='\t' || input=='\n';}
instead of the isBlankChar array?
```
because the function isBlankChar will be inlined by the compiler, also becuse there are only 3 tests and because the compiler can predict successfully the condition that will be satisfied?

Reply:
> Is the solution you are proposing having a function

Yes. But you're right, you can leave it for now until the character map is ready. It's not that important.

Reply:
> because the function isBlankChar will be inlined by the compiler
Yes
> also becuse there are only 3 tests and because the compiler can predict successfully the condition that will be satisfied?
No, the compiler doesn't predict. It may prove in certain cases. The one who predicts is the CPU.
That was the 3 sentence answer. I will follow with more extended info which anyone who doesn't care can skip.

**Extended Answer**:

-- About branch prediction --
To start simple, with prediction, we mean branch prediction. This is when you have some code like:
```cpp
if (condition) { ... }
```
When the code _runs_, the CPU has something called the branch predictor that for every branch (which is essentially an `if`) tries to predict whether it will be taken or not (i.e. whether you will execute the code inside or you'll skip it. Branches are really bad for modern CPUs (i.e. they slow things down) and so the CPU _predicts_ e.g. that in the next execution of a specific branch, the branch will be taken. So, it then starts executing the instructions "inside" the branch _without yet knowing if the branch should have been taken_ (i.e. the condition is not yet evaluated). Then, at some point, the condition is evaluated, and at this point the CPU may see that the branch should have not been taken. In that case, it has to undo the job it did (because it mistakenly took a branch and started executing its instructions). But, _most_ of the time, the CPU predicts correctly, and so this whole scheme makes the execution a lot faster (because the CPU does not wait for the evaluation of the condition to decide).

A totally naive branch predictor is an 1 bit branch predictor. i.e. if it was taken the last time,
assume that it will be taken now. Then, you can go to a 2-bit, which a little more complex but let's not bore you to death (not way more complex, you can google it).
One important thing to remember though is that the branch predictor in your chip
is _way_ (but I mean _wayyy_) crazier than that. It uses very complex machine learning algorithms and way more bits. You can check [that](https://youtu.be/hgcNM-6wr34?t=831) for a good intro, although you'll have to understand what is pipelining to get it.

-- About the role of the compiler --

Now, all this happens _at runtime_. The compiler can't predict what will happen at runtime (otherwise, we would do symbolic execution). The compiler can only:
1) Prove.
2) Optimize for certain cases based on heuristics.
3) Optimize on certain cases based on runtime.

1) Prove
Consider this:
```cpp
int test(int a) {
  if (a) return 1;
  return 2;
}

void foo() {
  int v = test(0);
}
```
The compiler here can (potentially) _prove_ (let's skip for a bit how it does it) that you won't take the branch (for this specific call, i.e. test(0)) and so the code is essentially transformed
to: `v = 2`;
But this is not based on the execution, but on something that is constant and known at compile-time.

2) Optimize for certain cases
The compiler may assume (using heuristics) that e.g. a specific virtual function is called
more frequently than the same (but other) virtual function. So, it may inline it
at the points it is called. This is again, something assumed and something not based
on runtime. Still, these heuristics are studied very much and so this more often than not
is actually an improvement (There's a whole concept related to this called
devirtualization).
You can check [this talk](https://youtu.be/3MRxucTXPdw?t=1352) (in the moment I picked, i.e. 22:30). It is the same concept through function pointers, which is essentially what virtual functions are.

3) Optimize based on runtime data.
This is something called PGO -> Profile-Guided Optimization, where in simple terms,
the compiler runs your code (having put instructions in it that gather data), gathers
data and then using these data finds that e.g. one function is called way too often so it inlines it.

-- About inlining --
Inling is a misunderstood optimization. Most people think that inlining is beneficial because it eliminates a call and a return to and from a function. That's not really the biggest benefit
in most cases. Inlining is mainly an _enabling_ optimization. That is, it mainly helps
_other_ optimizations. How does it do that? Well, by providing more context.
In the example above, if you inline `test()`, it's way easier to reason that v will take the value 2.
If it wasn't for inlining, you would have to do more sophisticated analysis, analysis that should take consideration code between different functions (there's a whole area for this, called Interprocedural Optimization or IPO).
Now, in this specific case, because your function is so small, the fact that we will eliminate a call and a return instruction will probably be beneficial as well.

I hope that was helpful and not very confusing.

Reply:
That was indeed really helpful. Really great answer.
Some follow up questions:

>This is something called PGO -> Profile-Guided Optimization, where in simple terms,
the compiler runs your code

What if the program when executed expects a user input or some arguments must be be specified from the command line? How will the compiler run the code?

Also in the talk about the function pointers the heuristic optimizations happen based on the profiler, so should't that be considered PGO ?

Reply:
> That was indeed really helpful. Really great answer.

Thanks, I'm glad it was helpful.

>> This is something called PGO -> Profile-Guided Optimization, where in simple terms,
>>  the compiler runs your code

> What if the program when executed expects a user input or some arguments must be be specified from the command line? How will the compiler run the code?

PGO is the idea. Essentially, optimization always implies that you from generic to specific. That is, for optimization to be effective, your data (that your program handles) have to be constrained in _some_ way. The idea is that you test inside this range (this is the very high level idea).

> Also in the talk about the function pointers the heuristic optimizations happen based on the profiler, so should't that be considered PGO ?

Yes, but you could skip that and e.g. imagine that the heuristics transform the code as in the talk based e.g. on how many times the function is called (I mean not on the runtime, but actually in how many places the function is called (which btw is not that simple if you take into consideration loops and recursion)).
For now, what matters is the idea and the idea is that we try to assume some things about the code and then based on that, we do transformations for the most common case (optimization always has to do about optimizing for the most common case).